BJS2-16351 Like system

fokinms opened the pull request:
Create LikeController with methods:
-likePost
-unlikePost
-likeComment
-unlikeComment

Create LikeDto & LikeMapper
Create LikeService
Create LikeServiceValidator with using FeignClient Create PostService
Create CommentService
Changed LikeRepository

fokinms committed:
Create LikeController with methods:
-likePost
-unlikePost
-likeComment
-unlikeComment

Create LikeDto & LikeMapper
Create LikeService
Create LikeServiceValidator with using FeignClient
Create PostService
Create CommentService
Changed LikeRepository

fraifree commented on src/main/java/faang/school/postservice/service/like/LikeService.java:
```diff
@@ -0,0 +1,66 @@
+    @Transactional
+    public LikeDto likePost(Long postId, Long userId) {
+        Post post = postService.findById(postId).orElseThrow(() -> {
+            log.error("Post with id {} does not exist", postId);
```
Лучше использовать log.info потому что отсутствие поста в базе никак не влияет на работу программы, его просто нет

fokinms replied:
Done. Случайно error написал)

fraifree commented on src/main/java/faang/school/postservice/service/like/LikeService.java:
```diff
@@ -0,0 +1,66 @@
+    @Transactional
+    public LikeDto likeComment(Long commentId, Long userId) {
+        Comment comment = commentService.findById(commentId).orElseThrow(() -> {
+            log.error("Comment with id {} does not exist", commentId);
```
same

fokinms replied:
Done

fokinms committed:
fix comments after cross-review

InnMish commented on src/main/java/faang/school/postservice/controller/like/LikeController.java:
```diff
@@ -0,0 +1,45 @@
+
+    @PostMapping("/user/{userId}/post/{postId}")
+    @ResponseStatus(HttpStatus.OK)
+    public LikeDto likePost(@Valid @PathVariable Long postId, @Valid @PathVariable Long userId) {
```
а что  @Valid на этом поле делает? 
на сколько я знаю - ничего) 
точно нужна эта аннотация здесь?

fokinms replied:
Done

InnMish commented on src/main/java/faang/school/postservice/controller/like/LikeController.java:
```diff
@@ -0,0 +1,45 @@
+
+    @DeleteMapping("/user/{userId}/post/{postId}")
+    @ResponseStatus(HttpStatus.OK)
+    public void unlikePost(@Valid @PathVariable Long postId, @Valid @PathVariable Long userId) {
```
valid надо?

InnMish commented on src/main/java/faang/school/postservice/service/like/LikeService.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Transactional
+    public LikeDto likePost(Long postId, Long userId) {
+        Post post = postService.findById(postId).orElseThrow(() -> {
+            log.info("Post with id {} does not exist", postId);
+            return new EntityNotFoundException("Post with id " + postId + " does not exist");
```
а пусть пост сервис эту ошибку кидает, если не найдена сущность 
это точно не  логика лайк сервиса - кидать ошибку про пост

fokinms replied:
Done

InnMish commented on src/main/java/faang/school/postservice/service/like/validation/LikeServiceValidator.java:
```diff
@@ -0,0 +1,65 @@
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class LikeServiceValidator {
```
LikeValidator просто

InnMish commented on src/main/java/faang/school/postservice/service/like/validation/LikeServiceValidator.java:
```diff
@@ -0,0 +1,65 @@
+package faang.school.postservice.service.like.validation;
```
пакет validator

InnMish commented on src/main/java/faang/school/postservice/service/like/validation/LikeServiceValidator.java:
```diff
@@ -0,0 +1,65 @@
+        } catch (FeignException e) {
+            if (e.status() == 500) {
+                log.info("User with id {} does not exist", userId);
+                throw new EntityNotFoundException("User with id " + userId + " does not exist");
```
а разве us не кинет сам такую ошибку? 
а 500 разве это EntityNotFound в нашем случае?

fokinms replied:
Done. UserService сам кинет, а тут пока убрал. Скорее всего надо будет отдельную логику для обработки такого типа ошибок прикрутить

InnMish commented on src/main/java/faang/school/postservice/service/like/validation/LikeServiceValidator.java:
```diff
@@ -0,0 +1,65 @@
+
+    public void validateByUserAndPostLikePossibility(Post post, Long userId) {
+        validateUser(userId);
+        validateLikingPostLogicalPossibility(post, userId);
```
название метода не говорит о том, что мы там собираемся делать 
очень размыто называется 
сейчас почитаю логику и предложу варианты

InnMish commented on src/main/java/faang/school/postservice/service/like/validation/LikeServiceValidator.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    private void validateLikingPostLogicalPossibility(Post post, Long userId) {
+        if (post.getLikes().stream().anyMatch(like -> like.getUserId().equals(userId))) {
```
хехе 
ну нет 
прикинь, ты хочешь лайкнуть пост, но у тебя лаганул интернет, и ты вообще лайкнул, но у тебя лайк не отобразился 
ты жмешь еще раз, а сервис тебе ошибку кидает, мол, ты уже лайкнул, куда гонишь)) 
я не думаю, что мы тут должны именно ошибку кинуть 
тут бы просто проверить, что мы еще не лайкали, но не кидать ошибку 
то есть метод сделать boolean 
если условие прошло - true, если нет - false

fokinms replied:
если прошло - ок, а если нет то чего делать?) ошибку же кинуть в итоге надо где-то, мол не могу лайк учесть. Либо если такой лайк уже стоит, то в сервисе тупо ОК возвращать? не очень понимаю как использовать здесь true/false

InnMish commented on src/main/java/faang/school/postservice/service/like/validation/LikeServiceValidator.java:
```diff
@@ -0,0 +1,65 @@
+        }
+        if (post.getComments().stream()
+                .flatMap(comment -> comment.getLikes().stream())
+                .anyMatch(like -> like.getUserId().equals(userId))) {
```
то же самое 
давай не будем кидать ошибку 

а вообще почему тут такая проверка? если я лайкнула какой-то коммент в посте, я не могу лайкнуть пост?

fokinms replied:
Ага. по заданию нельзя одновременно лайкать и пост, и комментарий относящийся к нему

InnMish commented on src/main/java/faang/school/postservice/service/like/validation/LikeServiceValidator.java:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    private void validateLikingPostLogicalPossibility(Post post, Long userId) {
```
validateNotLikedYet

fokinms replied:
Поменял. Чуть добавил по заданию

InnMish commented on src/main/java/faang/school/postservice/service/like/validation/LikeServiceValidator.java:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    private void validateLikingCommentLogicalPossibility(Comment comment, Long userId) {
```
тут по названию и логике внутри также, как выше

fokinms committed:
fix comments
Create DataValidationException
Create LikeServiceTest
Create LikeValidatorTest
change structure

Commit:
Merge branch 'refs/heads/griffon-master-bc5' into BJS2-16351

# Conflicts:
#	src/main/java/faang/school/postservice/service/CommentService.java
#	src/main/java/faang/school/postservice/service/PostService.java

Commit:
conflict solving

Commit:
tests fixed

Commit:
tests fixed

Commit:
tests fixed
